File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans=[]
        def dfs(i, total, collection):
            if total==target:
                ans.append(collection.copy())
                return
            if total>target or i>=len(candidates):
                return
            collection.append(candidates[i])
            dfs(i, total+candidates[i], collection)
            collection.pop()
            dfs(i+1, total, collection)
        
        dfs(0, 0, [])
        return ans

File: 39-combination-sum/39-combination-sum.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        ans=new ArrayList<>();
        combRec(candidates, 0, target, new ArrayList<>());
        return ans;
    }
    void combRec(int[] candidates, int ptr, int target, List<Integer> list){
        if(target==0){
            ans.add(new ArrayList<>(list));
            return;
        }
        if(target<0 || ptr>=candidates.length) return;
        list.add(candidates[ptr]);
        combRec(candidates, ptr, target-candidates[ptr], list);
        list.remove(list.size()-1);
        combRec(candidates, ptr+1, target, list);
    }
}